src/NodeVisitors/DefinitionCollector.php

```diff
@@ -0,0 +1,72 @@
+<?php
+declare(strict_types = 1);
+
+namespace LanguageServer\NodeVisitors;
+
+use PhpParser\{NodeVisitorAbstract, Node};
+
+/**
+ * Collects definitions of classes, interfaces, traits, methods, properties and constants
+ * Depends on ReferencesAdder and NameResolver
+ */
+class DefinitionCollector extends NodeVisitorAbstract
+{
+    /**
+     * Map from fully qualified name (FQN) to Node
+     * Examples of fully qualified names:
+     *  - testFunction()
+     *  - TestNamespace\TestClass
+     *  - TestNamespace\TestClass::TEST_CONSTANT
+     *  - TestNamespace\TestClass::staticTestProperty
+     *  - TestNamespace\TestClass::testProperty
+     *  - TestNamespace\TestClass::staticTestMethod()
+     *  - TestNamespace\TestClass::testMethod()
+     *
+     * @var Node[]
+     */
+    public $definitions = [];
+
+    public function enterNode(Node $node)
+    {
+        if ($node instanceof Node\Stmt\ClassLike && isset($node->name)) {
+            // Class, interface or trait declaration
+            $this->definitions[(string)$node->namespacedName] = $node;
```

Comment:
Fair enough, it just looked like a bug at first because you test for one attribute but access another.